Fix blanket `# noqa` in astropy

WilliamJamieson opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

tox -e codestyle

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This pull request fixes all of the blanket `# noqa` throughout `astropy`. Either by fixing the reason for the `# noqa`, removing it as it was unnecessary, or by adding the appropriate  violation code to be skipped.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #13702

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

WilliamJamieson commented:
@nstarman or @pllim should I break this into multiple PRs?

pllim commented:
For less than 300 lines, I think it can stay in one PR unless this drags on and you get a lot of conflicts, but I hope not.

nstarman commented on astropy/coordinates/tests/test_pickle.py:
```diff
@@ -64,7 +64,7 @@ def test_pickle_longitude_wrap_angle():
 @pytest.mark.parametrize(("name", "args", "kwargs", "xfail"),
                          tuple(zip(_names, _args, _kwargs, _xfail)))
-def test_simple_object(pickle_protocol, name, args, kwargs, xfail):
+def test_simple_object(pickle_protocol, name, args, kwargs, xfail):  # noqa: F811
```
It would be nice to ignore this noqa, but ONLY in tests. Pytest fixtures do not play nicely with PEP8.

nstarman replied:
Looking at https://github.com/pre-commit/pygrep-hooks/blob/main/.pre-commit-hooks.yaml, I don't think refinement is possible, so the noqa is fine.

nstarman commented on astropy/coordinates/tests/test_sky_coord.py:
```diff
@@ -360,8 +360,8 @@ def test_equal():
     assert np.all(ne == [False, True])
-    assert (sc1[0] == sc2[0]) == True  # noqa  (numpy True not Python True)
-    assert (sc1[0] != sc2[0]) == False  # noqa
+    assert sc1[0] == sc2[0]  # (numpy True not Python True)
```
No need for the comment anymore.

```suggestion
    assert sc1[0] == sc2[0]
```

nstarman commented on astropy/cosmology/__init__.py:
```diff
@@ -8,16 +8,16 @@
-from .parameter import *  # noqa F401, F403
-from .realizations import available, default_cosmology  # noqa F401, F403
-from .utils import *  # noqa F401, F403
+from .core import *  # noqa: F401, F403
```
Oops. I had the intent, but not the correct execution 😆

nstarman commented on astropy/table/tests/test_operations.py:
```diff
@@ -828,8 +830,8 @@ def _setup(self, t_cls=Table):
         out = table.setdiff(self.t1, self.t2)
-        assert type(out['a']) is type(self.t1['a'])  # noqa
-        assert type(out['b']) is type(self.t1['b'])  # noqa
+        assert isinstance(out['a'], type(self.t1['a']))
```
I'd restore a noqa to these types of tests. The ``isinstance`` replacement is less strict a comparison than the ``is``.

WilliamJamieson replied:
There are a lot more of these issues scattered throughout `astropy` (we don't catch them because we don't acutally run our entire `flake8` config).

So instead of adding a bunch more `# noqa` flags I added a helper function which implements the "type is type" assertion (strangely `flake8` is fine with that one?). I did this in fb658d8, do you think this is better?

nstarman replied:
I think so! Might need a changelog entry for it

WilliamJamieson replied:
Just added one!

nstarman commented on astropy/time/tests/test_basic.py:
```diff
@@ -2381,7 +2381,7 @@ def test_location_init(location):
     tm2 = Time(tm)
     assert np.all(tm.location == tm2.location)
-    assert type(tm.location) is type(tm2.location)  # noqa
+    assert isinstance(tm.location, type(tm2.location))
```
Same here.

nstarman reviewed (changes requested):
Thanks @WilliamJamieson for going through all these files!

nstarman commented on astropy/tests/helper.py:
```diff
@@ -470,3 +471,16 @@ def assert_quantity_allclose(actual, desired, rtol=1.e-7, atol=None,
+
+
+def assert_type_is_type(a, b):
+    """
+    Assert that two objects have exactly the same type.
+
+    If a or b is a type, then compare directly to that type instead.
+    """
```
This file is actually public-ish API, so a more detailed doctring would be great.

```suggestion
    """Assert that two objects have exactly the same type.

    Parameters
    ------------
    a, b : object or type
        The two objects to compare. If either is a `type`,
         then compare directly to that type instead.

    Raises
    -------
    AssertionError
        If the type of ``a`` is not the type of ``b``.

    Examples
    ----------
    The following pass:

        >>> assert_type_is_type(int, int)
        >>> assert_type_is_type(1, int)
    """
```

nstarman reviewed (approved):
Thanks @WilliamJamieson for all the work! LGTM. I'll let other weigh in...

pllim commented on astropy/coordinates/tests/test_frames.py:
```diff
@@ -890,8 +890,8 @@ def test_equal():
     assert np.all(ne == [False, True])
-    assert (sc1[0] == sc2[0]) == True  # noqa  (numpy True not Python True)
-    assert (sc1[0] != sc2[0]) == False  # noqa
+    assert sc1[0] == sc2[0]
+    assert not (sc1[0] != sc2[0])
```
I would like coordinates maintainers to have a look at this one: @StuartLittlefair, @adrn, @eteq, @eerovaher, @mhvk (according to Team page)

mhvk replied:
I think the point here was indeed to show that `__eq__` returns a boolean (and not something like `"boo!"`, which would evaluate like true...), so I'd leave this one as it was (adding the relevant warning code).

eerovaher replied:
If the intention is to test that the return values are indeed Booleans then the correct way to assert that would be
```python
assert (sc1[0] == sc2[0]) is True
assert (sc1[0] == sc2[0]) is False
```

mhvk replied:
Yes, but note the comment in the original line: one gets a numpy True, not a python one...

taldcroft replied:
Would it make sense to write `assert (sc1[0] == sc2[0]) is np.bool_(True)`? Then the test is really insisting on getting the numpy True.

eerovaher replied:
We really should implement @taldcroft's suggestion because the current test does not distinguish between Python and `numpy` True:
```python
>>> import numpy as np
>>> np.bool_(True) == True
True
```

mhvk replied:
The test is really just to show that `__eq__` returns something like a boolean (i.e., not `sc` or `"boo!"` that evaluates as true). I think it is an implementation detail whether that is a numpy or python bool, so would keep the test as it was originally. Or do `assert isinstance(sc1[0] == sc2[0], (bool, np.bool_))` -- that makes the intent clearer.

WilliamJamieson replied:
See 757d1d9fd8c8b55ade21f6dc2d566c9ca5e8047f

saimn replied:
While reviewing the black PR for io.fits I see this line 
```
        assert isinstance(v := (tbhdu.columns.columns[4].array[0] == np.True_), (bool, np.bool_)) and v
```

which brings me here through git blame, to ask ... whyyy ?? 
The original code was just testing that an array element was a True value. Why does it need to be so complicated just to avoid a `noqa` flag ?

WilliamJamieson replied:
@saimn, wen I originally "fixed" this, i attempted to change this to `assert tbhdu.columns.columns[4].array[0] is np.True_` (the `noqa` flag was suppressing the suggestion to change this). @mhvk then identified this as a test which was testing the result of the `__eq__` function; meaning that, I needed to change this back. Moreover, @mhvk noted that there are lots of return values from `__eq__` which will technically don't fail the `assert` such as returning `"boo!"`, which means we ought to test that the result is a boolean (the `bool` vs `np.bool_` type was an implementation detail). Hence the result turned into a test of the type of the equality return value.

After thinking about this, the original tests were essentially testing that the array elements were `True` (under the assumption that `__eq__` would be returning a boolean). Which means that the new `isinstance` check would not be able to detect if "True" and "False" were accidentally reversed in the `__eq__` evaluation (or some other pathological bug), which the previous iteration of the test could in fact detect  (assuming the result was a boolean). So I added the walrus assignment to test that the `__eq__` was also the expected outcome in addition to the right type. This was probably overkill, but it means the adjustments to the test should cover all the same cases as before in addition to confirming the boolean nature of the the result.

Feel free to change it back if you would like.

pllim commented on astropy/io/ascii/cds.py:
```diff
@@ -188,7 +188,7 @@ def process_lines(self, lines):
                       if x.startswith(('------', '======='))]
         if not i_sections:
             raise core.InconsistentTableError(f'No {self._subfmt} section delimiter found')
-        return lines[i_sections[-1]+1:]  # noqa
```
Are you sure it won't complain about the lack of space after colon?

WilliamJamieson replied:
With our current flake8 configuration (from setup.cfg), there are no complaints from that whole file.

eerovaher replied:
It would be better to ensure all these lines are `black`-compliant.

eerovaher replied:
More lines being `black`-compliant already means the diffs `black` will create will be smaller.

mhvk commented on astropy/coordinates/tests/test_sky_coord.py:
```diff
@@ -360,8 +360,8 @@ def test_equal():
     assert np.all(ne == [False, True])
-    assert (sc1[0] == sc2[0]) == True  # noqa  (numpy True not Python True)
-    assert (sc1[0] != sc2[0]) == False  # noqa
+    assert sc1[0] == sc2[0]
```
Same here

pllim commented on astropy/io/fits/tests/test_table.py:
```diff
@@ -1177,7 +1178,7 @@ def test_add_data_by_rows(self):
         assert (tbhdu.columns.columns[3].array[0] ==
                 np.array([0., 0., 0., 0., 0.], dtype=np.float32)).all()
-        assert tbhdu.columns.columns[4].array[0] == True  # noqa
+        assert tbhdu.columns.columns[4].array[0]
```
@saimn , is it ok to remove `== True` here?

pllim commented on astropy/table/tests/test_column.py:
```diff
@@ -796,7 +796,7 @@ def test_masked_col_unicode_sandwich():
 
     ok = c == ['abc', 'def']
-    assert ok[0] == True  # noqa
+    assert ok[0]
```
@taldcroft , is it okay to remove `== True` here?

taldcroft replied:
Yes.

mhvk commented on astropy/io/ascii/tests/test_ecsv.py:
```diff
@@ -965,7 +966,7 @@ def test_masked_vals_in_array_subtypes():
     for name in t2.colnames:
         assert t2[name].dtype == t[name].dtype
-        assert type(t2[name]) is type(t[name])  # noqa
+        assert_type_is_type(t2[name], t[name])
```
A bit sad to have to introduce a new function just to avoid code checkers!  I'd prefer the relevant `noqa`...

WilliamJamieson replied:
I undid this change in 757d1d9fd8c8b55ade21f6dc2d566c9ca5e8047f

mhvk commented on astropy/io/fits/hdu/image.py:
```diff
@@ -1082,7 +1082,7 @@ def match_header(cls, header):
         # Due to problems discussed in #5808, we cannot assume the 'GROUPS'
         # keyword to be True/False, have to check the value
         return (card.keyword == 'SIMPLE' and
-                ('GROUPS' not in header or header['GROUPS'] != True) and  # noqa
```
I think this is a functional change:
```
In [8]: 2 != True
Out[8]: True
```

mhvk commented on astropy/io/fits/tests/test_table.py:
```diff
@@ -1239,7 +1240,7 @@ def test_assign_multiple_rows_to_table(self):
         assert tbhdu2.columns.columns[2].array[0] == ''
         assert (tbhdu2.columns.columns[3].array[0] ==
                 np.array([0., 0., 0., 0., 0.], dtype=np.float32)).all()
-        assert tbhdu2.columns.columns[4].array[0] == True  # noqa
```
Again, this may not right, it may well have been important to test that the array holds things equivalent to a bool

WilliamJamieson replied:
Fixed by 757d1d9fd8c8b55ade21f6dc2d566c9ca5e8047f

pllim commented on astropy/tests/helper.py:
```diff
@@ -470,3 +471,33 @@ def assert_quantity_allclose(actual, desired, rtol=1.e-7, atol=None,
+        >>> assert_type_is_type(1, 2)
+    """
+
+    t_a = a if (t_a := type(a)) is type else t_a
```
And @mhvk for the usage in time tests.

mhvk commented on astropy/tests/helper.py:
```diff
@@ -470,3 +471,33 @@ def assert_quantity_allclose(actual, desired, rtol=1.e-7, atol=None,
         actual, desired, rtol, atol), **kwargs)
+
+
+def assert_type_is_type(a, b):
```
As mentioned above, I find it a bit sad to have a function whose only reason is to avoid style checkers. Maybe we just do not enforce this particular item?

taldcroft replied:
I think it is a relatively common mistake to check for types being equal instead of `isinstance`, so maybe leaving this check in place servers some value. If so then it seems we have two choices (1) use `# noqa (E721)` or (2) the new function. 

I'm a little worried about trying to be too clever here with the function as @mhvk pointed out. It's also going to be a bit obscure and it may be difficult to remember to use that. So maybe just using the explicit `noqa` in these cases will convey the intent most clearly?

Being "less clever" with the function as a drop-in replacement would essentially reduce it to `assert a is b`, which does seem a bit silly. With this code even the `# noqa` is not needed since the checker doesn't know `a` and `b` are types.

WilliamJamieson replied:
Based on @taldcroft's comment I decided to remove this and just use `# noqa: E721` instead.

mhvk commented on astropy/tests/helper.py:
```diff
@@ -470,3 +471,33 @@ def assert_quantity_allclose(actual, desired, rtol=1.e-7, atol=None,
+        >>> assert_type_is_type(1, 2)
+    """
+
+    t_a = a if (t_a := type(a)) is type else t_a
```
This is not quite good enough:
```
In [9]: type (Time)
Out[9]: abc.ABCMeta
```
I think you actually want `isinstance(type(a), type)` :smiley_cat:

pllim commented on astropy/units/tests/test_units.py:
```diff
@@ -171,7 +171,7 @@ def test_unknown_unit3():
     assert unit != u.m
     # next two from gh-7603.
-    assert unit != None  # noqa
+    assert unit is not None
```
Why not just slap the `# noqa: E711` to be consistent with the test below?

WilliamJamieson replied:
Fixed in 757d1d9fd8c8b55ade21f6dc2d566c9ca5e8047f

pllim commented on docs/changes/tests/13754.feature.rst:
```diff
@@ -0,0 +1,3 @@
+Add the ``assert_type_is_type`` test helper function which asserts that types,
```
Since this is exposed as "public" API here. Do we need a proper test to cover all the supported cases? The doctest snippet is pretty bare.

mhvk commented on astropy/units/tests/test_units.py:
```diff
@@ -171,7 +171,7 @@ def test_unknown_unit3():
     assert unit == "FOO"
     assert unit != u.m
     # next two from gh-7603.
-    assert unit != None  # noqa
```
I think here we are testing `unit.__ne__` so this should stay.

WilliamJamieson replied:
Fixed by 757d1d9fd8c8b55ade21f6dc2d566c9ca5e8047f

mhvk reviewed (commented):
Thanks for doing this fairly thankless task!

I think the comparisons with `True`, `False`, and `None` should all stay -- they are meant to test `__eq__`, `__ne__`.

I'm also not very charmed with adding a function just for the sake of avoiding style checkers; let's just put the relevant number there (or allow it more generally).

mhvk commented:
Seeing that @pllim and I had so many similar comments, perhaps the easiest way to resolve it is to make this PR one were we only make sure there are no bare `noqa`, either by removing them if they were not needed, or adding the specific warning.

Or do that in one commit and then the rest in separate ones (as already done for `assert_type_is_type` -- I think I'd drop that but can be convinced otherwise!)

pllim commented:
Would definitely be easier to approve and merge faster if there is no significant code change besides `#noqa` stuff...

WilliamJamieson commented:
Since these are all undocumented `# noqa` flags, it makes sense to attempt to fix whatever is causing the issue if possible.

Would it be better just to strip all the blanket `# noqa` flags instead.

astrofrog commented on astropy/constants/__init__.py:
```diff
@@ -23,11 +23,11 @@
+from . import si  # noqa: F401, E402
+from . import utils as _utils  # noqa: F401, E402
+from .config import codata, iaudata  # noqa: F401, E402
+from .constant import Constant, EMConstant  # noqa: F401, E402
```
I'm curious if there would be a better way to deal with the warnings in ``__init__.py`` files since this is so common. There's nothing to fix here, so I wonder if we can somehow run flake8 in a way that it would automatically ignore F401/E402 when looking at ``__init__.py`` files?

astrofrog commented on astropy/coordinates/tests/test_pickle.py:
```diff
@@ -6,9 +6,9 @@
 from astropy import coordinates as coord
 from astropy.coordinates import Longitude, representation
-from astropy.tests.helper import check_pickling_recovery, pickle_protocol  # noqa
+from astropy.tests.helper import check_pickling_recovery, pickle_protocol  # noqa: F401
```
Is this an unused import? Or is it because ``flake8`` doesn't understand that this is being used as a fixture?

WilliamJamieson replied:
This is one of those unfortunate cases of `pytest` not being very "pythonic", which means, as you suggested, `flake8` doesn't understand its being used.

astrofrog commented on astropy/coordinates/tests/test_sky_coord_velocities.py:
```diff
@@ -17,7 +17,7 @@
     UnitSphericalRepresentation)
 from astropy.tests.helper import assert_quantity_allclose
-from astropy.utils.compat.optional_deps import HAS_SCIPY  # noqa
+from astropy.utils.compat.optional_deps import HAS_SCIPY  # noqa: F401
```
Maybe cases like this mean that we should write ``skipif`` decorators using actual Python code instead of strings?

WilliamJamieson replied:
I think this would be better if it would work.

astrofrog reviewed (commented):
I wonder if it would be worth splitting this effort into three:

* Removing any ``noqa`` that are no longer needed (this can probably be automated to avoid manual work). This should then be pretty uncontroversial.
* Finding a way to deal specifically with unused import warnings in ``__init__.py`` files since it's pretty tedious to have to clutter those files with so many ``noqa``s and is such a common thing.
* The rest

nstarman commented:
> Finding a way to deal specifically with unused import warnings in __init__.py files since it's pretty tedious to have to clutter those files with so many noqas and is such a common thing.

My 2 cents... From my time spent static typing and compiling other libraries, I've found that star imports are better replaced by explicit imports. Pre-commit flake8 checks make sure the `__all__` in `__init__` stays synced with the module content.

WilliamJamieson commented:
> My 2 cents... From my time spent static typing and compiling other libraries, I've found that star imports are better replaced by explicit imports. Pre-commit flake8 checks make sure the `__all__` in `__init__` stays synced with the module content.

Alternately, we could add a [`per-file-ignores`](https://flake8.pycqa.org/en/latest/user/options.html#cmdoption-flake8-per-file-ignores) entry to our flake8 config to ignore F401/E402 in `__init__.py` files.

astrofrog commented:
I think @WilliamJamieson's suggestion would be best for now - and we can use wildcards so shouldn't be too hard.

WilliamJamieson commented:
I like @astrofrog's suggestions on how to break this PR apart, so I think I'll create two separate PRs with the first removing all the unnecessary `# noqa` and the second forcing `flake8` and `pre-commit` to ignore the F401/E402 errors in `__init__.py`.

I will leave this PR open and try to keep it up to date so that we can discuss the more controversial `# noqa` changes here.

WilliamJamieson commented:
I think I have addressed all the comments so far. I backed off on all the `type is type` changes and used `is np.bool_` in others.

I also removed all the `__init__.py` changes from this PR as those will be covered by #13781 when it merged. This means this will need to be rebased after #13781 is merged.

WilliamJamieson commented on astropy/io/fits/tests/test_table.py:
```diff
@@ -1239,7 +1239,7 @@ def test_assign_multiple_rows_to_table(self):
         assert (tbhdu2.columns.columns[3].array[0] ==
                 np.array([0., 0., 0., 0., 0.], dtype=np.float32)).all()
-        assert tbhdu2.columns.columns[4].array[0] == True  # noqa
+        assert tbhdu2.columns.columns[4].array[0] is np.bool_(True)
```
@mhvk do you think this is better, or would something akin to your suggestion here https://github.com/astropy/astropy/pull/13754#discussion_r983622791 be better?

mhvk replied:
This can just be `np.True_`.

But, yes, I like my suggestion perhaps a little better as it makes clear that the intent of the test is to verify that type of the output from `__eq__` -- and if `numpy` ever where to return python bool scalars, that would be fine!

jdavies-st commented:
> Pre-commit flake8 checks make sure the `__all__` in `__init__` stays synced with the module content.

Do note that `flake8` will not ensure that `__all__` is correct, only that there are no unused imports.  I.e. it won't report that the following bugs:

```python
__all__ = ['myfunction', 'myclass', myvariable]
```

or

```python
__all__ = {'myfunction', 'myclass', 'myvariable'}
```

`pylint` will find these, but it will also find lots of false positives as well.

Don't ask me how I know this.  😅

mhvk commented on astropy/table/tests/test_operations.py:
```diff
@@ -831,8 +831,8 @@ def _setup(self, t_cls=Table):
-        assert type(out['a']) is type(self.t1['a'])  # noqa
-        assert type(out['b']) is type(self.t1['b'])  # noqa
+        assert type(out['a']) is type(self.t1['a'])  # noqa: E721
+        assert type(out['a']) is type(self.t1['a'])  # noqa: E721
```
Note the typo: this should check column 'b'

mhvk commented on astropy/table/tests/test_operations.py:
```diff
@@ -842,8 +842,8 @@ def test_default_same_tables(self, operation_table_type):
-        assert type(out['a']) is type(self.t1['a'])  # noqa
-        assert type(out['b']) is type(self.t1['b'])  # noqa
+        assert type(out['a']) is type(self.t1['a'])  # noqa: E721
+        assert type(out['a']) is type(self.t1['a'])  # noqa: E721
```
Same

mhvk commented on astropy/table/tests/test_operations.py:
```diff
@@ -857,8 +857,8 @@ def test_extra_col_right_table(self, operation_table_type):
-        assert type(out['a']) is type(self.t1['a'])  # noqa
-        assert type(out['b']) is type(self.t1['b'])  # noqa
+        assert type(out['a']) is type(self.t1['a'])  # noqa: E721
+        assert type(out['a']) is type(self.t1['a'])  # noqa: E721
```
Same

mhvk commented on astropy/table/tests/test_operations.py:
```diff
@@ -868,8 +868,8 @@ def test_keys(self, operation_table_type):
-        assert type(out['a']) is type(self.t1['a'])  # noqa
-        assert type(out['b']) is type(self.t1['b'])  # noqa
+        assert type(out['a']) is type(self.t1['a'])  # noqa: E721
+        assert type(out['a']) is type(self.t1['a'])  # noqa: E721
```
Same

mhvk commented on astropy/table/tests/test_operations.py:
```diff
@@ -922,8 +922,8 @@ def test_stack_rows(self, operation_table_type):
-        assert type(out['a']) is type(self.t1['a'])  # noqa
-        assert type(out['b']) is type(self.t1['b'])  # noqa
+        assert type(out['a']) is type(self.t1['a'])  # noqa: E721
+        assert type(out['a']) is type(self.t1['a'])  # noqa: E721
```
Same

mhvk reviewed (commented):
Looks good! I found a few mistaken changes (typos), though.

mhvk reviewed (approved):
Looks all good! Thanks for sticking with this!

mhvk commented:
Failure due to a timeout, so unrelated
